File: templates/fsharp/xplat/AvaloniaTest/ViewLocator.fs
namespace AvaloniaTest

open System
open Avalonia.Controls
open Avalonia.Controls.Templates
open AvaloniaTest.ViewModels

type ViewLocator() =
    interface IDataTemplate with
        
        member this.Build(data) =
            if isNull data then
                null
            else    
                let name = data.GetType().FullName.Replace("ViewModel", "View", StringComparison.Ordinal)
                let typ = Type.GetType(name)
                if isNull typ then
                    upcast TextBlock(Text = sprintf "Not Found: %s" name)
                else
                    downcast Activator.CreateInstance(typ)

        member this.Match(data) = data :? ViewModelBase

File: templates/csharp/xplat/AvaloniaTest/ViewLocator.cs
using System;
using Avalonia.Controls;
using Avalonia.Controls.Templates;
using AvaloniaTest.ViewModels;

namespace AvaloniaTest;

public class ViewLocator : IDataTemplate
{
#if (AvaloniaStableChosen)
    public IControl Build(object param)
#else
    public Control? Build(object param)
#endif
    {
		if (param is null)
			return null;

		var name = param.GetType().FullName!.Replace("ViewModel", "View", StringComparison.Ordinal);
		var type = Type.GetType(name);

		if (type != null)
		{
			return (Control)Activator.CreateInstance(type)!;
		}

		return new TextBlock { Text = name };
	}

    public bool Match(object? data)
    {
        return data is ViewModelBase;
    }
}